aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,288 @@
+#pragma once
+
+// define constants like M_PI and C keywords for MSVC
+#ifdef _MSC_VER
+#define _USE_MATH_DEFINES
+#include <math.h>
+#endif
+
+#include <ATen/CPUGenerator.h>
+#include <ATen/core/Array.h>
+#include <type_traits>
+#include <limits>
+#include <cmath>
+
+/** 
+ * Distributions kernel adapted from THRandom.cpp
+ * The kernels try to follow std::random distributions signature
+ * For instance: in ATen
+ *      auto gen = at::detail::createCPUGenerator();
+ *      at::uniform_real_distribution<double> uniform(0, 1);
+ *      auto sample = uniform(gen.get());
+ *      
+ *      vs std::random
+ * 
+ *      std::mt19937 gen;
+ *      std::uniform_real_distribution uniform(0, 1);
+ *      auto sample = uniform(gen);
+ */
+
+
+namespace at {
+
+// Using VectorType in Box-muller derived distributions to avoid
+// code duplication
+template <typename T>
+struct VectorType {  };
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <> struct VectorType<half> { using type = at::detail::Array<float, 2>; };
+#endif
+template <> struct VectorType<Half> { using type = at::detail::Array<float, 2>; };
+template <> struct VectorType<float> { using type = at::detail::Array<float, 2>; };
+template <> struct VectorType<double> { using type = at::detail::Array<double, 2>; };
+
+template <typename T>
+using vect_type = typename VectorType<T>::type;
+
+// Using DistAccumType in accumulate types for distributions.
+// Note: Ideally we'd be using ATen/AccumulateType.h but looks
+// like the there is some inconsistency in how accumulate types
+// are mapped currently, e.g. for the cpu side, float is mapped
+// to double.
```

Comment:
@ngimel Do you know what's going on here?

Reply:
The accumulate type of float being double for the cpu side probably propagated from THNN: https://github.com/pytorch/pytorch/pull/6786.